### ISSUE TITLE
Handle configuration change using onSavedInstance

### ISSUE THREAD

**Issue by vrajdesai78** [oppia/oppia-android#4470]
**Problem**
For all interactions in exploration player and question player when we rotate our device, selected or entered answer doesn't persist. 

**Solution**
To solve this problem of configuration change we use onSavedInstance method to store the last entered user answer. 

**Alternatives**
- **Singleton Class Approach**: Complextiy of code will be increased if we use this approach as we have to write additional singleton class. Also, we need to handle issue of memory leak separately as it is not life-cycle aware. 
- **ViewModel Approach**: It is difficult to handle interactions when we have more than two interactions on one screen as all interactions are tightly coupled. 

Note: Closing issue #1737 as the approach for configuration change is changed to onSavedInstance from Singleton class approach.

**Comment by masclot** [comment]
The ui/presenter is subscribed to get EphemeraState updates [here](https://github.com/oppia/oppia-android/blob/324023bb727007572fd2ab233c2e106a39bfe9eb/app/src/main/java/org/oppia/android/app/player/state/StateFragmentPresenter.kt#L279). The subscription is not removed after first data comes, so theoretically new data can come at any time that recreates the ui. In this case, the user input is also lost. Should I handle this case as well or in practice it won't happen?

If it can happen and I need to handle the case, what changes can I expect? could the interaction id have changed? Even if the interaction id is the same, can it belong to a different question, in which case the saved data must be discarded?

Potential scenario I'm imagining:
- The student opens Oppia both in smartphone and pc/web
- Student opens the same exploration in both
- Student types something in both, but submits only one

Another potential case:
- The student opens Oppia android
- Student navigates to an exploration and starts typing and answer
- There is a change in the exploration configuration received over the net (Is this actually possible?)
- The new exploration configuration changes the interaction id from a text input to multiple choice
- The Ui/presenter receives the new exploration configuration and recreates the view
- The student input is lost

@BenHenning 
@adhiamboperes

**Comment by BenHenning** [comment]
@masclot the app is currently local only, so fortunately none of the network scenarios you've outlined can occur. Regarding stability, the actual source of truth is whatever is provided via ``EphemeralState`` (or question), and yes it can change from underneath you since that's what happens during lesson navigation. No assumptions can be made about its state during a configuration change, though we could theoretically eliminate that particular bit of uncertainty by saving/restoring the latest ephemeral state while the new fragment ways for its ``DataProvider`` subscription to callback (I hadn't thought of doing something like this before--it might actually lead to smoother configuration changes).

Anything that's transient UI state and not part of the ephemeral data structures will be lost on configuration change. User input/interaction selections are the most obvious that are affected by this. FWIW if there's an easy way represent partial interaction answers in the ephemeral structures, that might actually be an easier way to go than trying to make saved instance state work. It also potentially leads to a cleaner long-term solution of saving these protos on disk for low-memory death scenarios, as well (but that's not part of the scope of this issue so it doesn't specifically need to be considered here).

**Comment by adhiamboperes** [comment]
Unassigning @masclot because I feel this is blocked on technical direction. See https://github.com/oppia/oppia-android/pull/5008#issuecomment-1627516505
